dump/mapping/prepbufr_adpupa.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python3
+import os
+import numpy as np
+import time
+import calendar
+from datetime import datetime
+
+import bufr
+from bufr.obs_builder import ObsBuilder, add_main_functions, map_path
+from prepbufr_obs_builder import PrepbufrObsBuilder
+
+MAPPING_PATH = map_path('prepbufr_adpupa.yaml')
+
+
+class AdpupaPrepbufrObsBuilder(PrepbufrObsBuilder):
+    """
+    A builder class to generate ADPUPA observations from ADPUPA prepBUFR input.
+    """
+
+    def __init__(self):
+        super().__init__(MAPPING_PATH, log_name=os.path.basename(__file__))
+
+    def make_obs(self, comm, input_path):
+
+        # Get container from mapping file first
+        self.log.info(f'Get container from bufr')
+        container = super().make_obs(comm, input_path)
+
+        self.log.debug(f'container list (original): {container.list()}')
+
+        self.log.debug(f'Perform DateTime calculation')
+        hrdr = container.get('obsTimeMinusCycleTime')
+        cycle_time = "2021080100"
+        cycleTimeSinceEpoch = np.int64(calendar.timegm(time.strptime(str(int(cycle_time)), '%Y%m%d%H')))
+        dateTime = self._compute_datetime(cycleTimeSinceEpoch, hrdr)
+
+        self.log.debug(f'Perform stationPressure calculation')
+        cat = container.get('prepbufrDataLevelCategory')
+        pob = container.get('pressure')
+        ps = np.full(pob.shape[0], pob.fill_value)
+        ps = np.where(cat == 0, pob, ps)
```

Comment:
could just change the `mask` attribute. Ex:

```python
station_pressure = container.get('pressure')
station_pressure.mask = station_pressure.mask && (cat == 0)
```

Reply:
@rmclaren done, thanks! Spelled the variables out and added a new function for this to aviod repetition of the codes.